### PR TITLE
[Feature] Add `extract_subcoco` script

### PR DESCRIPTION
**Pull request by hhaAndroid**
## Motivation

The training dataset of the COCO2017 dataset includes 118k images, and the validation set includes 5k images, which is a relatively large dataset. Loading JSON in debugging or quick verification scenarios will consume more resources and bring slower startup speed.
The `extract_subcoco.py` script provides the ability to extract a specified number of images. The user can use the `-num-img` parameter to get a COCO subset of the specified number of images.

Currently, only support COCO2017. In the future will support user-defined datasets of standard coco JSON format.

The root path folder format is as follows:

```text
├── root
│   ├── annotations
│   ├── train2017
│   ├── val2017
│   ├── test2017
```

1. Extract 10 training images and 10 validation images using only 5k validation sets.

```shell
python tools/misc/extract_subcoco.py ${ROOT} ${OUT_DIR} --num-img 10
```

2. Extract 20 training images using the training set and 20 validation images using the validation set.

```shell
python tools/misc/extract_subcoco.py ${ROOT} ${OUT_DIR} --num-img 20 --use-training-set
```

### CODE REVIEW

File: docs/en/user_guides/useful_tools.md [hunk 1]
```diff
@@ -183,3 +183,32 @@ python tools/model_converters/yolox_to_mmyolo.py --src yolox_s.pth --dst mmyolox
 ```
 
 The converted `mmyolox.pt` can be used by MMYOLO.
+
+## Extracts a subset of COCO
+
+The training dataset of the COCO2017 dataset includes 118k images, and the validation set includes 5k images, which is a relatively large dataset. Loading JSON in debugging or quick verification scenarios will consume more resources and bring slower startup speed.
+The `extract_subcoco.py` script provides the ability to extract a specified number of images. The user can use the `-num-img` parameter to get a COCO subset of the specified number of images.
```
Comment (PeterH0323) [inline docs/en/user_guides/useful_tools.md:190]:
```suggestion
The `extract_subcoco.py` script provides the ability to extract a specified number of images. The user can use the `--num-img` parameter to get a COCO subset of the specified number of images.
```

File: docs/en/user_guides/useful_tools.md [hunk 2]
```diff
@@ -183,3 +183,32 @@ python tools/model_converters/yolox_to_mmyolo.py --src yolox_s.pth --dst mmyolox
...
+Currently, only support COCO2017. In the future will support user-defined datasets of standard coco JSON format.
+
+The root path folder format is as follows:
+
+```text
+├── root
+│   ├── annotations
+│   ├── train2017
+│   ├── val2017
+│   ├── test2017
```
Comment (PeterH0323) [inline docs/en/user_guides/useful_tools.md:201]:
```suggestion
├── root
│   ├── annotations
│   ├── train2017
│   ├── val2017
│   └── test2017
└ ...
```

File: docs/en/user_guides/useful_tools.md [hunk 3]
```diff
@@ -183,3 +183,32 @@ python tools/model_converters/yolox_to_mmyolo.py --src yolox_s.pth --dst mmyolox
 ```
 
 The converted `mmyolox.pt` can be used by MMYOLO.
+
+## Extracts a subset of COCO
+
+The training dataset of the COCO2017 dataset includes 118k images, and the validation set includes 5k images, which is a relatively large dataset. Loading JSON in debugging or quick verification scenarios will consume more resources and bring slower startup speed.
```
Comment (PeterH0323) [inline docs/en/user_guides/useful_tools.md:189]:
```suggestion
The training dataset of the COCO2017 dataset includes 118K images, and the validation set includes 5K images, which is a relatively large dataset. Loading JSON in debugging or quick verification scenarios will consume more resources and bring slower startup speed.
```

File: docs/en/user_guides/useful_tools.md [hunk 4]
```diff
@@ -183,3 +183,32 @@ python tools/model_converters/yolox_to_mmyolo.py --src yolox_s.pth --dst mmyolox
...
+
+```text
+├── root
+│   ├── annotations
+│   ├── train2017
+│   ├── val2017
+│   ├── test2017
+```
+
+1. Extract 10 training images and 10 validation images using only 5k validation sets.
```
Comment (PeterH0323) [inline docs/en/user_guides/useful_tools.md:204]:
```suggestion
1. Extract 10 training images and 10 validation images using only 5K validation sets.
```

File: docs/zh_cn/user_guides/useful_tools.md [hunk 5]
```diff
@@ -181,3 +181,32 @@ python tools/model_converters/yolox_to_mmyolo.py --src yolox_s.pth --dst mmyolox
 ```
 
 转换好的 `mmyolox.pt` 即可以在 MMYOLO 中使用。
+
+## 提取 COCO 子集
+
+COCO2017 数据集训练数据集包括 118k 张图片，验证集包括 5k 张图片，数据集比较大。在调试或者快速验证程序是否正确的场景下加载 json 会需要消耗较多资源和带来较慢的启动速度，这会导致程序体验不好。
```
Comment (PeterH0323) [inline docs/zh_cn/user_guides/useful_tools.md:187]:
```suggestion
COCO2017 数据集训练数据集包括 118K 张图片，验证集包括 5K 张图片，数据集比较大。在调试或者快速验证程序是否正确的场景下加载 JSON 会需要消耗较多资源和带来较慢的启动速度，这会导致程序体验不好。
```

File: docs/zh_cn/user_guides/useful_tools.md [hunk 6]
```diff
@@ -181,3 +181,32 @@ python tools/model_converters/yolox_to_mmyolo.py --src yolox_s.pth --dst mmyolox
...
+注意： 本脚本目前仅仅支持 COCO2017 数据集，未来会支持更加通用的 COCO JSON 格式数据集
+
+输入 root 根路径文件夹格式如下所示：
+
+```text
+├── root
+│   ├── annotations
+│   ├── train2017
+│   ├── val2017
+│   ├── test2017
```
Comment (PeterH0323) [inline docs/zh_cn/user_guides/useful_tools.md:199]:
```suggestion
├── root
│   ├── annotations
│   ├── train2017
│   ├── val2017
│   └── test2017
└ ...
```

File: docs/zh_cn/user_guides/useful_tools.md [hunk 7]
```diff
@@ -181,3 +181,32 @@ python tools/model_converters/yolox_to_mmyolo.py --src yolox_s.pth --dst mmyolox
...
+```
+
+1. 仅仅使用 5k 张验证集切分出 10 张训练图片和 10 张验证图片
```
Comment (PeterH0323) [inline docs/zh_cn/user_guides/useful_tools.md:202]:
```suggestion
1. 仅仅使用 5K 张验证集切分出 10 张训练图片和 10 张验证图片
```

File: tools/misc/extract_subcoco.py [hunk 8]
```diff
@@ -0,0 +1,119 @@
...
+
+This script is mainly used to debug and verify the correctness of the
+program quickly.
+The root folder format must be in the following format:
+
+├── root
+│   ├── annotations
+│   ├── train2017
+│   ├── val2017
+│   ├── test2017
```
Comment (PeterH0323) [inline tools/misc/extract_subcoco.py:12]:
```suggestion
│   └── test2017
└ ...
```

File: tools/misc/extract_subcoco.py [hunk 9]
```diff
@@ -0,0 +1,119 @@
...
+import os.path as osp
+import shutil
+
+import mmengine
+import numpy as np
+from pycocotools.coco import COCO
+
+
+# TODO: Currently only supports coco2017
+def _process_data(args, in_dataset_type: str, out_dataset_type: str):
```
Comment (PeterH0323) [inline tools/misc/extract_subcoco.py:31]:
```suggestion
def _process_data(args, in_dataset_type: str, out_dataset_type: str, year: str = '2017'):
```

File: tools/misc/extract_subcoco.py [hunk 10]
```diff
@@ -0,0 +1,119 @@
...
+    assert in_dataset_type in ('train', 'val')
+    assert out_dataset_type in ('train', 'val')
+
+    year = '2017'
```
Comment (PeterH0323) [inline tools/misc/extract_subcoco.py:35]:
```suggestion
   assert year == '2017', 'only support COCO2017'
```

File: tools/misc/extract_subcoco.py [hunk 11]
```diff
@@ -0,0 +1,119 @@
...
+        'categories': json_data['categories'],
+        'images': [],
+        'annotations': []
+    }
+
+    images = json_data['images']
+    coco = COCO(ann_path)
+
+    # shuffle
+    np.random.shuffle(images)
```
Comment (PeterH0323) [inline tools/misc/extract_subcoco.py:55]: 建议加入 seed 设置，默认 seed = -1，这里判断如果是  -1 的话就自动生成，如果非 -1 则设置，最后打印 seed 到 logger

File: tools/misc/extract_subcoco.py [hunk 12]
```diff
@@ -0,0 +1,119 @@
...
+    _make_dirs(args.out_dir)
+
+    print('start processing train dataset')
+    if args.use_training_set:
+        _process_data(args, 'train', 'train')
+    else:
+        _process_data(args, 'val', 'train')
+    print('start processing val dataset')
+    _process_data(args, 'val', 'val')
+
```
Comment (PeterH0323) [inline tools/misc/extract_subcoco.py:116]: 建议加上 log 打印保存的根目录，并显示 all done 完成的字样

File: tools/misc/extract_subcoco.py (hunk 11 above)
Comment (hhaAndroid) [reply tools/misc/extract_subcoco.py:55]: 没有设置就不设置了吧

Comment (PeterH0323) [reply tools/misc/extract_subcoco.py:55]: 是的

**Review by PeterH0323 — APPROVED** [review]
LGTM

**Review by RangeKing — APPROVED** [review]
